internal/moduleref/resolver.go

```diff
@@ -44,40 +45,55 @@ func NewResolver(internalManifest modsdir.Manifest) *Resolver {
 // and return a new manifest encapsulating this information.
 func (r *Resolver) Resolve(cfg *configs.Config) *Manifest {
 	// First find all the referenced modules.
-	r.findAndTrimReferencedEntries(cfg)
+	r.findAndTrimReferencedEntries(cfg, nil, nil)
 
 	return r.manifest
 }
 
 // findAndTrimReferencedEntries will traverse a given Terraform configuration
 // and attempt find a caller for every entry in the internal module manifest.
 // If an entry is found, it will be removed from the internal manifest and
-// appended to the manifest that records this new information.
-func (r *Resolver) findAndTrimReferencedEntries(cfg *configs.Config) {
-	for entryKey, entry := range r.internalManifest {
-		for callerKey := range cfg.Module.ModuleCalls {
-			// Construct the module path with the caller key to get
-			// the full module entry key. If it's a root module caller
-			// do nothing since the path will be empty.
-			path := strings.Join(cfg.Path, ".")
-			if path != "" {
-				callerKey = path + "." + callerKey
+// appended to the manifest that records this new information in a nested heirarchy.
+func (r *Resolver) findAndTrimReferencedEntries(cfg *configs.Config, parentRecord *Record, parentKey *string) {
+	var name string
+	var versionConstraints version.Constraints
+	if parentKey != nil {
```

Comment:
`parentKey` does not need to be a pointer, we can safely check if the string is empty since no module except the root module will have an empty key.